symfony/plugins/orangehrmLeavePlugin/lib/service/LeaveRequestService.php

```diff
@@ -573,10 +573,17 @@ public function changeLeaveStatus($changes, $changeType, $changeComments = null,
                     //LeaveEvents::LEAVE_CANCEL
 
                     foreach ($changedItems as $leaveRequestId) {
-                        $changedLeave = $this->searchLeave($leaveRequestId);
-                        $this->_changeLeaveStatus($changedLeave, $nextState, $changeComments[$leaveRequestId]);
-                        $this->_notifyLeaveStatusChange($event, $workFlow, $changedLeave, 
-                                $changedByUserType, $changedUserId, 'request');                        
+                        $changedLeaveRequest = $this->fetchLeaveRequest($leaveRequestId);
+                        $changedLeave = $changedLeaveRequest->getLeave();
+                        $allowedActions = $this->getLeaveRequestActions($changedLeaveRequest, $changedUserId);
+
+                        if (isset($allowedActions[$workFlowId])) {
+                            $this->_changeLeaveStatus($changedLeave, $nextState, $changeComments[$leaveRequestId]);
+                            $this->_notifyLeaveStatusChange($event, $workFlow, $changedLeave,
+                                $changedByUserType, $changedUserId, 'request');
+                        } else {
+                            throw new Exception('Action Not allowed');
```

Comment:
If we throw the exception in the middle of changing the leave status that may cause to half approved/canceled and others not